### PR TITLE
feat(text): add text recolor back

### PR DESCRIPTION
**Pull request by C47D**
Implements #6944

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

**Comment by C47D** [comment]
Missing:
- [ ] Update `lv_text_get_width`, it has multiple usages.
- [ ] Update `lv_label` widget

**Comment by C47D** [comment]
Current state with the following code:
image

```c
lv_label_set_text(label, "Write a #ff0000 red# word");
```

**Comment by kisvegabor** [comment]
Great progress! Let me know when it's ready for review!

**Comment by C47D** [comment]
> Great progress! Let me know when it's ready for review!

Hi @kisvegabor , could you help me with this questions?
- Update lv_text_get_width, it has multiple usages (OR skip calling it in lv_text_get_size when the recolor flag is found)
- Update Kconfig, where should the new option be placed?

**Comment by C47D** [comment]
This is the failing rendering test, is there a way to see the failure? 

label_decor_err

**Comment by C47D** [comment]
> Hi @kisvegabor , could you help me with this questions?
> 
>     * Update `lv_text_get_width`, it has multiple usages (OR skip calling it in lv_text_get_size when the recolor flag is found)

This is a list of places where `lv_text_get_width` is being used, all of them are `inside` widgets, but I'm worried that it will break user code if we add the `flags` parameter to the function signature:
image

**Comment by C47D** [comment]
> This is the failing rendering test, is there a way to see the failure? Is the right side the actual render and the left side the expected render? 😅
> 
> ```
> /home/runner/work/lvgl/lvgl/tests/src/test_cases/draw/test_draw_label.c:187:test_label_decor:INFO: 
> Screenshot compare error
>   - File: ref_imgs/draw/label_decor.png
>   - Mode: full refresh
>   - At x:100, y:69.
>   - Expected: 0x00000000004C4AC9
>   - Actual:   0x0000000000E9A45B
>   - Tolerance: 0
> /home/runner/work/lvgl/lvgl/tests/src/test_cases/draw/test_draw_label.c:123:test_label_decor:FAIL: draw/label_decor.png
> ```
> 
> label_draw.png: label_decor
> 
> label_draw_error.png: label_decor_err

If I'm interpreting this correctly the selection color of the font is not kept in the label_draw_err file, so I guess that's the cause of the failing render test.

**Comment by C47D** [comment]
@kisvegabor When the label style selection color is set and we also have a recolor command in the same letters, which color should prevail?

**Comment by C47D** [comment]
> > This is the failing rendering test, is there a way to see the failure? Is the right side the actual render and the left side the expected render? 😅
> > ```
> > /home/runner/work/lvgl/lvgl/tests/src/test_cases/draw/test_draw_label.c:187:test_label_decor:INFO: 
> > Screenshot compare error
> >   - File: ref_imgs/draw/label_decor.png
> >   - Mode: full refresh
> >   - At x:100, y:69.
> >   - Expected: 0x00000000004C4AC9
> >   - Actual:   0x0000000000E9A45B
> >   - Tolerance: 0
> > /home/runner/work/lvgl/lvgl/tests/src/test_cases/draw/test_draw_label.c:123:test_label_decor:FAIL: draw/label_decor.png
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > label_draw.png: label_decor
> > label_draw_error.png: label_decor_err
> 
> If I'm interpreting this correctly the selection color of the font is not kept in the label_draw_err file, so I guess that's the cause of the failing render test.

Fixed, I was always overwriting the letter color, which was overwriting the selection color for letters inside the selection range.

**Comment by C47D** [comment]
In summary @kisvegabor , the open points are:

> > Hi @kisvegabor , could you help me with this questions?
> > ```
> > * Update `lv_text_get_width`, it has multiple usages (OR skip calling it in lv_text_get_size when the recolor flag is found)
> > ```
> 
> This is a list of places where `lv_text_get_width` is being used, all of them are inside widgets, but I'm worried that adding the `flags` parameter at the end of the function signature will break user code if they're using it in their applications: image

---

> @kisvegabor When the label style selection color is set and we also have a recolor command in the same letters, which color should prevail?

---

When running the tests I'm hitting a `AddressSanitizer:DEADLYSIGNAL` error, but this also happens in master, maybe is my setup?
---

And:
Update Kconfig, should we also update lv_conf_kconf,h?

**Comment by kisvegabor** [comment]
> Update lv_text_get_width, it has multiple usages (OR skip calling it in lv_text_get_size when the recolor flag is found)

`lv_text_get_width` is really a public API and we should carefully change it. A new function could be called `lv_text_get_width_flags`, but it's a little bit odd. Anyway, in lack of better idea let's add `lv_text_get_width_flags`.

> When the label style selection color is set and we also have a recolor command in the same letters, which color should prevail?

Recolor can change only the letter color. So the selection color should stay as it is. 

> Update Kconfig, should we also update lv_conf_kconf,h?

`lv_conf_kconf,h` is just a helper for Kconfig. It seems we don't need to update it for this feature. 

> When running the tests I'm hitting a AddressSanitizer:DEADLYSIGNAL error, but this also happens in master, maybe is my setup?

Could you copy a longer long log?

**Comment by C47D** [comment]
This is the error I'm having:

```console
==============================
Running tests for test_defheap
==============================
Test project /home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/build_test_defheap
    Start  4: test_draw_label
    Start 73: test_label
1/2 Test #73: test_label .......................***Exception: SegFault  0.11 sec
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL
AddressSanitizer:DEADLYSIGNAL

2/2 Test  #4: test_draw_label ..................   Passed    0.17 sec

50% tests passed, 1 tests failed out of 2

Total Test time (real) =   0.17 sec

The following tests FAILED:
	 73 - test_label (SEGFAULT)
Errors while running CTest
```

And this is what gdb tells me when running that bin
```console
Reading symbols from build_test_defheap/test_label...
(gdb) r
Starting program: /home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/build_test_defheap/test_label 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:38:test_label_creation:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:44:test_label_set_text:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:57:test_label_get_letter_pos_align_left:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:85:test_label_get_letter_pos_align_left_on_empty_text:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:113:test_label_long_text_multiline_get_letter_pos_align_left:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:152:test_label_long_text_get_letter_pos_align_left:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:184:test_label_get_letter_pos_align_right:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:212:test_label_get_letter_pos_align_right_on_empty_text:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:240:test_label_long_text_multiline_get_letter_pos_align_right:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:279:test_label_long_text_get_letter_pos_align_right:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:311:test_label_get_letter_pos_align_center:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:339:test_label_get_letter_pos_align_center_on_empty_text:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:367:test_label_long_text_multiline_get_letter_pos_align_center:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:406:test_label_long_text_get_letter_pos_align_center:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:438:test_label_is_char_under_pos_align_left:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:461:test_label_is_char_under_pos_align_right:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:484:test_label_is_char_under_pos_align_center:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:507:test_label_cut_text:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:516:test_label_get_letter_on_left:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:529:test_label_get_letter_on_center:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:542:test_label_get_letter_on_right:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:555:test_label_text_selection:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:570:test_label_rtl_dot_long_mode:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:594:test_label_max_width:PASS
/home/d3bug/Desktop/lv_port_pc_eclipse/lvgl/tests/src/test_cases/widgets/test_label.c:618:test_label_with_recolor_cmd:PASS

-----------------------
25 Tests 0 Failures 0 Ignored 
OK
==27048==LeakSanitizer has encountered a fatal error.
==27048==HINT: For debugging, try setting environment variable LSAN_OPTIONS=verbosity=1:log_threads=1
==27048==HINT: LeakSanitizer does not work under ptrace (strace, gdb, etc)
[Inferior 1 (process 27048) exited with code 01]
(gdb) 
```

**Comment by C47D** [comment]
> Recolor can change only the letter color. So the selection color should stay as it is.

I think we can set the selected text color with this property `lv_obj_set_style_text_color(label, sel_color, LV_PART_SELECTED);` or maybe I'm confused :sweat_smile:

**Comment by C47D** [comment]
Hi,

I tried to implement avoiding counting when the recolor command is found but I wasn't able to figure out how to do it, I would like to propose to merge this as is. I'll update the docs once you think this is finished.

**Comment by egonbeermat** [comment]
I just gave this PR a whirl. I have 17 labels throughout my app using recolor, most of them simple one liners with a couple of colors, plus 2 larger text areas spanning several lines and colors. I'm happy to report all of them worked as expected, as per LVGL 8.x.x. 

Screenshots from this PR:
Weather app. Blue wind and rain glyphs are part of a single recolor label.


Checklist is text area text:


My in-app cheat sheet about all my settings, memory alloc, etc, one large text area with multiple colors:


Thanks for doing this!!

**Comment by C47D** [comment]
Thanks for reporting your successful test @egonbeermat , nice UI you got.

**Review by kisvegabor — DISMISSED** [review]
Looks good to me! 

I thinks a lot of people who were angry at me because of removing this feature will be very grateful for you. :slightly_smiling_face:

File: src/draw/lv_draw_label.c [hunk 1]
```diff
@@ -378,6 +437,49 @@ void lv_draw_label_iterate_characters(lv_draw_unit_t * draw_unit, const lv_draw_
...
+                break;
+            default:
+                result = 0;
+                break;
+        }
+    }
+
+    return result;
+}
+
```
Comment (W-Mai) [inline src/draw/lv_draw_label.c:482]:
```suggestion
/**
 * Convert a hexadecimal characters to a number (0..15)
 * @param hex Pointer to a hexadecimal character (0..9, A..F)
 * @return the numerical value of `hex` or 0 on error
 */
static uint8_t hex_char_to_num(char hex)
{
    if(hex >= '0' && hex <= '9') return hex - '0';
    if(hex >= 'a') hex -= 'a' - 'A'; /*Convert to upper case*/
    return 'A' <= hex && hex <= 'F' ? hex - 'A' + 10 : 0;
}
```

File: src/misc/lv_text.c [hunk 2]
```diff
@@ -369,6 +414,45 @@ int32_t lv_text_get_width(const char * txt, uint32_t length, const lv_font_t * f
...
+int32_t lv_text_get_width_with_flags(const char * txt, uint32_t length, const lv_font_t * font, int32_t letter_space,
+                                     lv_text_flag_t flags)
+{
+    if(txt == NULL) return 0;
+    if(font == NULL) return 0;
+    if(txt[0] == '\0') return 0;
+
+    uint32_t i                   = 0;
+    int32_t width             = 0;
+    lv_text_cmd_state_t cmd_state = LV_TEXT_CMD_STATE_WAIT;
```
Comment (W-Mai) [inline src/misc/lv_text.c:426]:
```suggestion
    uint32_t i                    = 0;
    int32_t width                 = 0;
    lv_text_cmd_state_t cmd_state = LV_TEXT_CMD_STATE_WAIT;
```

File: src/draw/lv_draw_label.c (hunk 1 above)
Comment (W-Mai) [inline src/draw/lv_draw_label.c]:
```suggestion
static uint8_t hex_char_to_num(char hex)
{
    if(hex >= '0' && hex <= '9') return hex - '0';
    if(hex >= 'a') hex -= 'a' - 'A'; /*Convert to upper case*/
    return 'A' <= hex && hex <= 'F' ? hex - 'A' + 10 : 0;
}
```

**Review by W-Mai — CHANGES_REQUESTED** [review]

File: src/misc/lv_text.c [hunk 3]
```diff
@@ -136,6 +136,37 @@ void lv_text_get_size(lv_point_t * size_res, const char * text, const lv_font_t
         size_res->y -= line_space;
 }
 
+bool lv_text_is_cmd(lv_text_cmd_state_t * state, uint32_t c)
```
Comment (W-Mai) [inline src/misc/lv_text.c:139]: the first params of a widget's public function need to be the obj it self.

I think you can get the state from lv_label_t casted from lv_obj_t.

Comment (C47D) [reply src/misc/lv_text.c:139]: Hi, thanks for the feedback, I checked the other `lv_text` interfaces (in `lv_text.h`) and none of them has the `lv_obj_t` as first parameter, maybe this interface is an exception? What do you think @kisvegabor

Comment (kisvegabor) [reply src/misc/lv_text.c:139]: lv_text is not for widgets, so we can't pass obj parameters.

**Comment by XuNeo** [comment]
@qinshijing @xaowang96 Please try if it fits your needs too. Thank you.

**Review by kisvegabor — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Comment by finger563** [comment]
This doesn't appear to be in the latest `9.2.2` release, is there a planned release soon that contains this?

**Comment by AndreCostaaa** [comment]
Hi @finger563. Yes, v9.3 will be released in less than 2 weeks

**Pull request merged**